### PR TITLE
Reapply the compiler flags change but separate out the 'language server' specific settings

### PR DESCRIPTION
**Pull request by rchiodo**
This is a reapplication of https://github.com/microsoft/pyright/pull/8704 but with also taking into account flags that have to always be set even when a pyrightconfig.json is present because these flags are not settable by a pyrightconfig.json.

I tested vscode-pyright with this change and it takes the diagnostic mode into account correctly now and watches for file changes. This wasn't working with the original submission.

Addresses https://github.com/microsoft/pyright/issues/8719

### CODE REVIEW

**Review by erictraut — APPROVED** [review]
Looks good

File: packages/pyright-internal/src/common/commandLineOptions.ts [hunk 1]
```diff
@@ -145,14 +136,35 @@ export class CommandLineOptions {
...
     disableTaggedHints?: boolean;
+}
+
+// Some options can be specified from a source other than the pyright config file.
+// This can be from command-line parameters or some other settings mechanism, like
+// that provided through a language client like the VS Code editor. These options
+// are later combined with those from the config file to produce the final configuration.
+export class CommandLineOptions {
+    // Settings that are possible to set in a config.json file.
+    configSettings: CommandLineConfigOptions = new CommandLineConfigOptions();
```
Comment (rchiodo) [inline packages/pyright-internal/src/common/commandLineOptions.ts:149]: I'm not super happy with this name. config is used so much here. 'environmentSettings' maybe? As that's what the configuration.md calls them mostly now. Although it also includes type checking settings :(

Comment (erictraut) [reply packages/pyright-internal/src/common/commandLineOptions.ts:149]: Hmm, "configSettings" seems good to me.

Comment (rchiodo) [reply packages/pyright-internal/src/common/commandLineOptions.ts:149]: typeCheckingSettings?

Comment (rchiodo) [reply packages/pyright-internal/src/common/commandLineOptions.ts:149]: Okay I'll just leave it.

**Pull request merged**